crypto/rand/rand_lib.c

```diff
@@ -25,6 +25,45 @@ static CRYPTO_RWLOCK *rand_meth_lock;
 static const RAND_METHOD *default_RAND_meth;
 static CRYPTO_ONCE rand_init = CRYPTO_ONCE_STATIC_INIT;
 
+#ifdef OPENSSL_RAND_SEED_RDTSC
+/*
+ * Since we get some randomness from the low-order bits of the
+ * high-speec clock, it can help.  But don't return a status since
+ * it's not sufficient to indicate whether or not the seeding was
+ * done.
+ */
+void rand_rdtsc(void)
+{
+    unsigned char c;
+    int i;
+
+    for (i = 0; i < 10; i++) {
+        c = (unsigned char)(OPENSSL_rdtsc() & 0xFF);
+        RAND_add(&c, 1, 0.5);
+    }
+}
+#endif
+
+#ifdef OPENSSL_RAND_SEED_RDRAND
+size_t OPENSSL_ia32_rdrand(void);
+extern unsigned int OPENSSL_ia32cap_P[];
+
+int rand_rdrand(void)
+{
+    size_t i, s;
+
+    if ((OPENSSL_ia32cap_P[1] & (1 << (62 - 32))) == 0)
+        return 0;
+
+    for (i = 0; i < RANDOMNESS_NEEDED; i += sizeof(s)) {
+        s = OPENSSL_ia32_rdrand();
```

Comment:
The RDSEED instruction is a better choice if it is available.  RESEED is intended to produce seed material for PRNGs, RDRAND is a PRNG itself.

```
if rdseed is available
    use rdseed
else if rdrand is available
    use rdrand
else
    do nothing
end if
```